File: resources/scripts/shot_spec.py
# -*- coding: utf-8 -*-
"""
Shot Spec（分镜师工单）共享契约 —— Python 端（C0 契约）。
与 TS 端 `src/shared/contracts/shotSpec.ts` **逐字段一一对应**（字段清单同源，杜绝双端漂移）。

职责：
- 定义 ShotSpec dataclass（S3 剪辑改道输入的解析目标）+ 各枚举常量；
- `validate_shot_spec()`：结构校验（schema + 枚举 + 段号合法 + subjects∈段人物），供 S2 校验与 S3 消费共用；
- 任何新增字段必须在这里与 TS 端同步修改（TODO C0-3：改为生成式单真源）。

对齐方案：§24.2（2026-09-18）。
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional

# ---------------- 枚举常量（与 TS 端 SHOT_SPEC_ENUMS 逐项一致） ----------------

MODE_NEW_SHOT = "NEW_SHOT"
MODE_CONTINUE_PREV = "CONTINUE_PREV"
SHOT_SPEC_MODE = [MODE_NEW_SHOT, MODE_CONTINUE_PREV]

SPATIAL_INDOOR_RESIDENCE = "INDOOR_RESIDENCE"
SPATIAL_INDOOR_PUBLIC = "INDOOR_PUBLIC"
SPATIAL_OUTDOOR_STREET = "OUTDOOR_STREET"
SPATIAL_OUTDOOR_NATURE = "OUTDOOR_NATURE"
SPATIAL_VEHICLE = "VEHICLE"
SPATIAL_TRANSIT_HUB = "TRANSIT_HUB"
SPATIAL_UNKNOWN = "UNKNOWN"
SPATIAL_TYPE = [
    SPATIAL_INDOOR_RESIDENCE, SPATIAL_INDOOR_PUBLIC, SPATIAL_OUTDOOR_STREET,
    SPATIAL_OUTDOOR_NATURE, SPATIAL_VEHICLE, SPATIAL_TRANSIT_HUB, SPATIAL_UNKNOWN,
]

PREF_EXTREME_LONG = "EXTREME_LONG"
PREF_LONG_SHOT = "LONG_SHOT"
PREF_FULL_SHOT = "FULL_SHOT"
PREF_MEDIUM_SHOT = "MEDIUM_SHOT"
PREF_MEDIUM_CLOSE = "MEDIUM_CLOSE"
PREF_CLOSE_SHOT = "CLOSE_SHOT"
PREF_EXTREME_CLOSE = "EXTREME_CLOSE"
PREFERRED_SHOT = [
    PREF_EXTREME_LONG, PREF_LONG_SHOT, PREF_FULL_SHOT,
    PREF_MEDIUM_SHOT, PREF_MEDIUM_CLOSE, PREF_CLOSE_SHOT, PREF_EXTREME_CLOSE,
]

CAM_STATIC = "STATIC"
CAM_PAN = "PAN"
CAM_TILT = "TILT"
CAM_PUSH = "PUSH"
CAM_PULL = "PULL"
CAM_FOLLOW = "FOLLOW"
CAMERA_DYNAMIC = [CAM_STATIC, CAM_PAN, CAM_TILT, CAM_PUSH, CAM_PULL, CAM_FOLLOW]

EMO_WARM = "WARM"
EMO_TENSE = "TENSE"
EMO_JOYFUL = "JOYFUL"
EMO_SAD = "SAD"
EMO_NEUTRAL = "NEUTRAL"
SHOT_EMOTION = [EMO_WARM, EMO_TENSE, EMO_JOYFUL, EMO_SAD, EMO_NEUTRAL]

AUDIO_NARRATION = "narration"
AUDIO_ORIGINAL = "original"
AUDIO_MODE = [AUDIO_NARRATION, AUDIO_ORIGINAL]

# 降级链 Level：值越小越严（§24.4）
L0, L1, L2, L3, L4, L5 = 0, 1, 2, 3, 4, 5
FALLBACK_LEVEL = [L0, L1, L2, L3, L4, L5]


@dataclass
class ShotSpec:
    """分镜师工单：一份对应「一个完整句」（sentence 档 matchUnitId）。字段对齐 §24.2。"""
    matchUnitId: str                       # 来源完整句主键（溯源用）
    mode: str                              # NEW_SHOT | CONTINUE_PREV
    segmentId: int                         # 剧情场次段号（圈定候选池）
    spatialType: str                       # 空间类型（门禁）
    targetSubjects: List[str] = field(default_factory=list)  # 主体（取自段 characters 白名单）
    actionType: Optional[str] = None       # 动作类别（可空）
    keyProp: Optional[str] = None          # 关键道具类别（可空）
    preferredShot: str = PREF_MEDIUM_SHOT  # 首选景别（软排）
    cameraDynamic: str = CAM_STATIC        # 运镜（软排）
    emotion: str = EMO_NEUTRAL             # 情绪（软排）
    audioMode: str = AUDIO_NARRATION       # narration | original
    fallbackLevel: int = L3                # 允许降到第几级
    atmosphereNote: Optional[str] = None   # 散文备注（仅软排/兜底）


def validate_shot_spec(spec) -> Dict[str, str]:
    """
    结构校验：枚举 + 段号合法 + 必要字段齐全。
    返回错误字典 {字段: 错误信息}；空字典表示通过。
    注意：subjects∈段人物 属「配场存疑校验」，需带段上下文件另行判定，不在此处（§24.11-F）。

    参数:
        spec: ShotSpec 实例或 dict
    返回:
        结构错误映射；为空即通过
    """
    errors: Dict[str, str] = {}
    scope = spec.__dict__ if isinstance(spec, ShotSpec) else (spec or {})

    def _get(k: str):
        return scope.get(k) if isinstance(scope, dict) else getattr(spec, str(k), None)

    if not _get("matchUnitId"):
        errors["matchUnitId"] = "missing"
    mode = _get("mode")
    if mode not in SHOT_SPEC_MODE:
        errors["mode"] = f"invalid:{mode}"
    seg = _get("segmentId")
    if not isinstance(seg, int) or seg < 1:
        errors["segmentId"] = f"invalid:{seg}"
    if _get("spatialType") not in SPATIAL_TYPE:
        errors["spatialType"] = f"invalid:{_get('spatialType')}"
    if _get("preferredShot") not in PREFERRED_SHOT:
        errors["preferredShot"] = f"invalid:{_get('preferredShot')}"
    if _get("cameraDynamic") not in CAMERA_DYNAMIC:
        errors["cameraDynamic"] = f"invalid:{_get('cameraDynamic')}"
    if _get("emotion") not in SHOT_EMOTION:
        errors["emotion"] = f"invalid:{_get('emotion')}"
    if _get("audioMode") not in AUDIO_MODE:
        errors["audioMode"] = f"invalid:{_get('audioMode')}"
    if _get("fallbackLevel") not in FALLBACK_LEVEL:
        errors["fallbackLevel"] = f"invalid:{_get('fallbackLevel')}"
    return errors

File: src/shared/contracts/shotSpec.ts
/**
 * Shot Spec（分镜师工单）共享契约——S2 分镜师 Agent 输出 / S3 剪辑改道输入的**唯一同源 schema**（§24.2）。
 *
 * 责任边界：
 * - 本文件是 C0 契约的 TS 唯一真源，字段严格对齐方案 §24.2；
 * - Python 端 `resources/scripts/shot_spec.py` 由本文件字段清单同步维护（后改为生成式，见 TODO C0-3）；
 * - 任何新增字段必须先改这里，再同步 Python，保证「一个完整句 = 一条工单」的双端同构。
 */

/* ==================== 枚举常量（与 Python 端逐项一一对应） ==================== */

/** 工单模式：有实体句开新镜；评论/过渡/抒情句顺延上一镜（不检索） */
export const ShotSpecMode = {
  NEW_SHOT: 'NEW_SHOT',
  CONTINUE_PREV: 'CONTINUE_PREV',
} as const;
export type ShotSpecModeValue = typeof ShotSpecMode[keyof typeof ShotSpecMode];

/** 空间类型（门禁枚举，词表来自 S1 场记单的 20 个地点类；不引第二套空间枚举） */
export const SpatialType = {
  INDOOR_RESIDENCE: 'INDOOR_RESIDENCE', // 室内·住宅
  INDOOR_PUBLIC: 'INDOOR_PUBLIC',       // 室内·公共场所
  OUTDOOR_STREET: 'OUTDOOR_STREET',     // 室外·街道
  OUTDOOR_NATURE: 'OUTDOOR_NATURE',     // 室外·自然景
  VEHICLE: 'VEHICLE',                   // 载具内部
  TRANSIT_HUB: 'TRANSIT_HUB',           // 交通枢纽
  UNKNOWN: 'UNKNOWN',                   // 泛化/未知类（S1 泛化类 6.8%）
} as const;
export type SpatialTypeValue = typeof SpatialType[keyof typeof SpatialType];

/** 景别（软排，只能取自该段实际存在的景别） */
export const PreferredShot = {
  EXTREME_LONG: 'EXTREME_LONG', // 全景·大全景
  LONG_SHOT: 'LONG_SHOT',       // 全景
  FULL_SHOT: 'FULL_SHOT',       // 中全景
  MEDIUM_SHOT: 'MEDIUM_SHOT',   // 中景
  MEDIUM_CLOSE: 'MEDIUM_CLOSE', // 中近景
  CLOSE_SHOT: 'CLOSE_SHOT',     // 近景
  EXTREME_CLOSE: 'EXTREME_CLOSE', // 特写
} as const;
export type PreferredShotValue = typeof PreferredShot[keyof typeof PreferredShot];

/** 运镜（软排） */
export const CameraDynamic = {
  STATIC: 'STATIC', // 固定
  PAN: 'PAN',       // 横摇
  TILT: 'TILT',     // 竖摇
  PUSH: 'PUSH',     // 推
  PULL: 'PULL',     // 拉
  FOLLOW: 'FOLLOW', // 跟
} as const;
export type CameraDynamicValue = typeof CameraDynamic[keyof typeof CameraDynamic];

/** 情绪（软排；链条无情绪字段时跳过，不罚） */
export const ShotEmotion = {
  WARM: 'WARM',         // 温馨
  TENSE: 'TENSE',       // 紧张
  JOYFUL: 'JOYFUL',     // 欢快
  SAD: 'SAD',           // 悲伤
  NEUTRAL: 'NEUTRAL',   // 中性
} as const;
export type ShotEmotionValue = typeof ShotEmotion[keyof typeof ShotEmotion];

/** 音频模式：narration=解说合成；original=原声段（须锁说话人、且不参与变速） */
export const AudioMode = {
  NARRATION: 'narration',
  ORIGINAL: 'original',
} as const;
export type AudioModeValue = typeof AudioMode[keyof typeof AudioMode];

/**
 * 降级链 Level（值越小越严）：L0-L5 全集见 §24.4。
 * - L0 硬门禁全命中（最严）
 * - L1..L4 逐级放宽（同段环境/空镜/同段任意未用等）
 * - L5 段内复用（额度 ≤2）
 * 实指句末级**不用** CONTINUE_PREV（防画面死锁）。
 */
export const FallbackLevel = {
  L0: 0,
  L1: 1,
  L2: 2,
  L3: 3,
  L4: 4,
  L5: 5,
} as const;
export type FallbackLevelValue = typeof FallbackLevel[keyof typeof FallbackLevel];

/* ==================== 工单本体 ==================== */

/**
 * Shot Spec 工单：@24.2-2154 逐字段对齐。
 * 一份工单对应「一个完整句」（sentence 档 matchUnitId），由 S2 分镜师 Agent 产出。
 */
export interface ShotSpec {
  /** 唯一主键：来源完整句的 matchUnitId（与步骤3/步骤5 同源），便于溯源 */
  matchUnitId: string;
  /** 每母句一句，若某句不需要开新镜则为 CONTINUE_PREV */
  mode: ShotSpecModeValue;
  /** 剧情场次段号（来自场记单 §24.10），圈定局部候选池 */
  segmentId: number;
  /** 空间类型（门禁）；高纯度段 locPurity≥0.6 时硬卡，<0.6 转软排 */
  spatialType: SpatialTypeValue;
  /** 主体（只能取自该段的 characters 白名单） */
  targetSubjects: string[];
  /** 动作类别（可空：纯环境句） */
  actionType?: string;
  /** 关键道具类别（可空） */
  keyProp?: string;
  /** 首选景别（软排；只能取自该段实际存在景别） */
  preferredShot: PreferredShotValue;
  /** 运镜（软排） */
  cameraDynamic: CameraDynamicValue;
  /** 情绪（软排） */
  emotion: ShotEmotionValue;
  /** 音频模式 */
  audioMode: AudioModeValue;
  /** 允许降到第几级（§24.4 降级链） */
  fallbackLevel: FallbackLevelValue;
  /** 散文备注：仅作软排序微调/兜底，不参与硬门禁 */
  atmosphereNote?: string;
  /** 🎬 展示透传（非开单依据，不开给 LLM）：本工单对应完整句的文案原文 */
  text?: string;
  /** 🎬 展示透传（非开单依据，不开给 LLM）：本工单对应完整句的画面意图描述 */
  visualIntent?: string;
}

/** 校验时合法枚举的集合（供结构校验 ① 与 JSON schema 共用） */
export const SHOT_SPEC_ENUMS = {
  mode: Object.values(ShotSpecMode),
  spatialType: Object.values(SpatialType),
  preferredShot: Object.values(PreferredShot),
  cameraDynamic: Object.values(CameraDynamic),
  emotion: Object.values(ShotEmotion),
  audioMode: Object.values(AudioMode),
  fallbackLevel: Object.values(FallbackLevel),
} as const;